tcl/register.tcl

```diff
@@ -2,46 +2,53 @@ namespace eval qc {
-    namespace eval ::${method} {}
-    set proc_name "::${method}::$path"
-    {*}[list proc $proc_name $proc_args $proc_body]
-
-    # Separate arg names and default values
-    set args {}
-    set defaults {}
-    foreach arg $proc_args {
-        if {[llength $arg] == 2} {
-            lappend args [lindex $arg 0]
-            dict set defaults [lindex $arg 0] [lindex $arg 1]
-        } else {
-            lappend args $arg
-        }
+    if { [llength $args] < 2 || [llength $args] > 4 } {
+        return -code error "Usage: qc::register method path ?args? ?body?"
     }
 
-    # Update the handlers nsv array.
-    qc::nsv_dict set handlers $method $path proc_name $proc_name
-    qc::nsv_dict set handlers $method $path args $args
-    qc::nsv_dict set handlers $method $path body $proc_body
-    qc::nsv_dict set handlers $method $path defaults $defaults    
-
-    if {$method eq "GET"} {
-        set http_methods [list HEAD GET]
+    set method [string toupper [lindex $args 0]]
+    set path [lindex $args 1]
+    # Add to registered nsv array
+    if { [qc::nsv_dict exists registered $method] } {
+        set patterns [qc::nsv_dict get registered $method]
+        if { [lsearch -exact $patterns $path] < 0 } {
+            # path hasn't been registered already
+            qc::nsv_dict lappend registered $method $path
+        }
     } else {
-        set http_methods [list "POST"]
+        qc::nsv_dict set registered $method $path
     }
 
-    # Add path to the filters database if not present.
-    foreach http_method $http_methods {
-        if { ! [qc::nsv_dict exists filters filter_validate $http_method $path] } {
-            qc::nsv_dict set filters filter_validate $http_method $path 1
-        }
-        if { ! [qc::nsv_dict exists filters filter_authenticate $http_method $path] } {
-            qc::nsv_dict set filters filter_authenticate $http_method $path 1
+    if { [llength $args] >= 3 } {
+        set proc_args [lindex $args 2]
+        # Separate arg names and default values
+        set arg_names {}
+        set defaults {}
+        foreach arg $proc_args {
+            if {[llength $arg] == 2} {
+                lappend arg_names [lindex $arg 0]
+                dict set defaults [lindex $arg 0] [lindex $arg 1]
+            } else {
+                lappend arg_names $arg
+            }
         }
     }
+
+    if { [llength $args] == 4 } {
+        set proc_body [lindex $args 3]
+        # Create the proc
+        namespace eval ::${method} {}
+        set proc_name "::${method}::$path"
+        {*}[list proc $proc_name $proc_args $proc_body]
+
+        # Update the handlers nsv array.
+        qc::nsv_dict set handlers $method $path proc_name $proc_name
+        qc::nsv_dict set handlers $method $path args $arg_names
+        qc::nsv_dict set handlers $method $path body $proc_body
+        qc::nsv_dict set handlers $method $path defaults $defaults  
+    }
 }
 
 proc qc::validate {method path proc_args proc_body} {
```

Comment:
In `#|` comment terminology "extra validation" or "custom validation"
`#| Register a URL handler for extra validation.`
`#| Register a URL handler for custom validation.`